BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,328 @@ def browse
+      format.json { render :json => api_response }
+      format.json { render :xml  => api_response }
     end
-    to_operate = newly_registered_userfiles - collisions
+  end
 
-    if collisions.size > 0
-      flash[:error] += "Could not #{move_or_copy.downcase} some files, as files with the same names already exist:\n" +
-                         collisions.map(&:name).sort.join(", ")
+  # Directly delete a list of files from a browsable DP
+  def delete
+    # Extract key parameters & make sure the provider is browsable
+    @provider = DataProvider.find_accessible_by_user(params[:id], current_user)
+    @as_user  = browse_as(params['as_user_id'])
+    unless @provider.is_browsable?(current_user)
+      flash[:error] = "You cannot delete files from this provider."
+      respond_to do |format|
+        format.html { redirect_to :action => :index }
+        format.xml  { render :xml  => { :error => flash[:error] }, :status => :forbidden }
+        format.json { render :json => { :error => flash[:error] }, :status => :forbidden }
+      end
+      return
     end
 
-    if to_operate.size == 0
-      flash[:error] += "No files are left to #{move_or_copy.downcase} !\n"
-    else
-      flash[:notice] += "Warning! #{to_operate.size} files are now being #{past_tense} in background.\n"
-      success_list  = []
-      failed_list   = {}
-      CBRAIN.spawn_with_active_records(:admin, "#{move_or_copy} Registered Files") do
-        to_operate.shuffle.each_with_index do |u,idx|
-          $0="#{move_or_copy} Registered Files ID=#{u.id} #{idx+1}/#{to_operate.size}\0\0\0\0"
-          begin
-            if move_or_copy == "MOVE"
-              u.provider_move_to_otherprovider(new_dp)
-              u.set_size!
-            elsif move_or_copy == "COPY" # and no ELSE !
-              new = u.provider_copy_to_otherprovider(new_dp)
-              u.delete rescue true # NOT destroy()! We don't want to delete the content!
-              u.destroy_log rescue true
-              new.set_size!
+    flash[:notice] ||= ''
+
+    # Erase the given userfiles in background.
+    userfiles = userfiles_from_basenames(@provider, @as_user, params[:basenames])
+    succeeded, failed = [], {}
+
+    CBRAIN.spawn_with_active_records_if(
+      [:html, :js].include?(request.format.to_sym),
+      current_user,
+      "Delete files (data_provider: #{@provider.id})"
+    ) do
+      userfiles.shuffle.each do |basename, userfile|
+        begin
+          # Is the userfile registered?
+          if userfile.present?
+            # Make sure the current user can delete the file
+            unless userfile.has_owner_access?(current_user)
+              (failed["Insufficient permissions"] ||= []) << basename
+              next
             end
-            success_list << u
-          rescue => ex
-            (failed_list[ex.message] ||= []) << u
+
+            result = userfile.destroy
+            userfile.destroy_log rescue true
```

Comment:
Here the destroy_log() is always done anyway if destroy() succeeds. No need to do it explicitely.